nixos/modules/services/backup/borgbackup.nix

```diff
@@ -0,0 +1,488 @@
+      exit_status=$?
+      # Reset the EXIT handler, or else we're called again on 'exit' below
+      trap - EXIT
+      ${cfg.postHook}
+      exit $exit_status
+    }
+    trap 'on_exit' INT TERM QUIT EXIT
+
+    ${cfg.preHook}
+  '' + optionalString cfg.doInit ''
+    # Run borg init if the repo doesn't exist yet
+    if ! borg list; then
+      ${optionalString (isPath cfg.repo) "mkdir -p ${escapeShellArg cfg.repo}"}
+      borg init \
+        --encryption ${cfg.encryption.mode} \
+        ${cfg.extraInitArgs}
+      ${cfg.postInit}
+    fi
+  '' + ''
+    borg create \
+      --compression ${cfg.compression} \
+      --exclude-from ${mkExcludeFile cfg} \
+      ${cfg.extraCreateArgs} \
+      ${mkArchiveArg cfg} \
+      ${mkPathsArgs cfg}
+    ${cfg.postCreate}
+  '' + optionalString (cfg.prune.keep != { }) ''
+    borg prune \
+      ${mkKeepArgs cfg} \
+      ${cfg.extraPruneArgs}
+    ${cfg.postPrune}
+  '' + ''
+    ${cfg.postHook}
+  '';
+
+  mkPassEnv = cfg: with cfg.encryption;
+    if passCommand != null then
+      { BORG_PASSCOMMAND = passCommand; }
+    else if passphrase != null then
+      { BORG_PASSPHRASE = passphrase; }
+    else { };
+
+  mkBackupService = name: cfg: 
+    let
+      userHome = config.users.users.${cfg.user}.home;
+    in nameValuePair "borgbackup-job-${name}" {
+      description = "BorgBackup job ${name}";
+      path = with pkgs; [
+        borgbackup openssh
+      ];
+      script = mkBackupScript cfg;
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        # Only run when no other process is using CPU or disk
+        CPUSchedulingPolicy = "idle";
+        IOSchedulingClass = "idle";
+        ProtectSystem = "full";
+        #FIXME:
+        #ProtectSystem = "strict";
+        #ReadWritePaths =
+        #  [ "${userHome}/.config/borg" "${userHome}/.cache/borg" ]
+        #  # Borg needs write access to repo if it is not remote
+        #  ++ optional (isPath cfg.repo) cfg.repo;
+      };
+      environment = {
+        BORG_REPO = cfg.repo;
+      } // (mkPassEnv cfg) // cfg.environment;
+      inherit (cfg) startAt;
+    };
+
+  # The name arg is needed for mapAttrsToList
+  mkPassAssertion = name: cfg: {
+    assertion = with cfg.encryption;
+      mode != "none" -> passCommand != null || passphrase != null;
+    message =
+      ''passCommand or passphrase have to be specified ''
+      + ''because borgbackup.jobs.${name}.encryption != "none"'';
+  };
+
+  mkRepoService = name: cfg:
+    nameValuePair "borgbackup-repo-${name}" {
+      description = "Create BorgBackup repository ${name} directory";
+      script = ''
+        mkdir -p ${escapeShellArg cfg.path}
+        chown ${cfg.user}:${cfg.group} ${escapeShellArg cfg.path}
+      '';
+      serviceConfig = {
+        # The service's only task is to ensure that the specified path exists
+        Type = "oneshot";
+      };
+      wantedBy = [ "multi-user.target" ];
+      after = [ "multi-user.service" ];
```

Comment:
Is this "after" needed? (I don't see this pattern anywhere in nixpkgs.)

Reply:
It isn't. Thanks for catching that.